[21:35:34 - main - INFO] Waiting for JIT...
Traced<ShapedArray(float64[])>with<DynamicJaxprTrace(level=3/1)>
Traced<ShapedArray(float64[])>with<DynamicJaxprTrace(level=3/1)>
Traced<ShapedArray(float64[])>with<DynamicJaxprTrace(level=3/1)>
Traced<ShapedArray(float64[])>with<DynamicJaxprTrace(level=3/1)>
Traced<ShapedArray(float64[])>with<DynamicJaxprTrace(level=3/1)>
Traced<ShapedArray(float64[])>with<DynamicJaxprTrace(level=3/1)>
Traced<ShapedArray(float64[])>with<DynamicJaxprTrace(level=3/1)>
0.7050593286633925
0.7309436708102671
0.4319256886395569
0.5165955259554958
0.5668696913378383
0.674474897052168
0.7681713637643213
[21:35:46 - main - INFO] ---------------------------
[21:35:46 - main - INFO] Iter:   0    Time: 11.246
[21:35:46 - main - INFO] -----------  --------------
[21:35:46 - main - INFO] ics_loss     1.757e+00
[21:35:46 - main - INFO] res_loss     9.885e+01
[21:35:46 - main - INFO] l2_error     1.544e+00
Traced<ShapedArray(float64[])>with<DynamicJaxprTrace(level=3/1)>
Traced<ShapedArray(float64[])>with<DynamicJaxprTrace(level=3/1)>
Traced<ShapedArray(float64[])>with<DynamicJaxprTrace(level=3/1)>
Traced<ShapedArray(float64[])>with<DynamicJaxprTrace(level=3/1)>
Traced<ShapedArray(float64[])>with<DynamicJaxprTrace(level=3/1)>
Traced<ShapedArray(float64[])>with<DynamicJaxprTrace(level=3/1)>
Traced<ShapedArray(float64[])>with<DynamicJaxprTrace(level=3/1)>
I1012 21:35:48.779325 8441426496 checkpoints.py:570] Saving checkpoint at step: 2
I1012 21:35:48.779988 8441426496 checkpoints.py:658] Using Orbax as backend to save Flax checkpoints. For potential troubleshooting see: https://flax.readthedocs.io/en/latest/guides/training_techniques/use_checkpointing.html#orbax-as-backend-troubleshooting
I1012 21:35:48.780106 8441426496 abstract_checkpointer.py:35] orbax-checkpoint version: 0.6.4
I1012 21:35:48.780328 8441426496 checkpointer.py:189] [process=0] Started saving checkpoint to /Users/ananyaebhartari/Documents/Paris/My_pirate/jaxpi/examples/allen_cahn/default/ckpt/checkpoint_2.
I1012 21:35:48.780809 8441426496 utils.py:240] [process=0][thread=MainThread] Skipping global process sync, barrier name: create_tmp_directory:pre.checkpoint_2.1
I1012 21:35:48.781177 8441426496 atomicity.py:175] Creating tmp directory /Users/ananyaebhartari/Documents/Paris/My_pirate/jaxpi/examples/allen_cahn/default/ckpt/checkpoint_2.orbax-checkpoint-tmp-0
I1012 21:35:48.781841 8441426496 checkpoint.py:134] Wrote CheckpointMetadata=CheckpointMetadata(init_timestamp_nsecs=1728783348781540000, commit_timestamp_nsecs=None), json={"init_timestamp_nsecs": 1728783348781540000, "commit_timestamp_nsecs": null} to /Users/ananyaebhartari/Documents/Paris/My_pirate/jaxpi/examples/allen_cahn/default/ckpt/checkpoint_2.orbax-checkpoint-tmp-0
I1012 21:35:48.781975 8441426496 utils.py:240] [process=0][thread=MainThread] Skipping global process sync, barrier name: create_tmp_directory:post.checkpoint_2.2
I1012 21:35:48.789175 8441426496 base_pytree_checkpoint_handler.py:108] [process=0] /jax/checkpoint/write/blocking_bytes_per_sec: 1.3 GiB/s (total bytes: 9.1 MiB) (time elapsed: 6 milliseconds) (per-host)
I1012 21:35:48.809406 8441426496 future.py:88] ChainedFuture completed 2/2 futures in 0.02 seconds.
I1012 21:35:48.809525 8441426496 base_pytree_checkpoint_handler.py:108] [process=0] /jax/checkpoint/write/bytes_per_sec: 336.4 MiB/s (total bytes: 9.1 MiB) (time elapsed: 27 milliseconds) (per-host)
I1012 21:35:48.809755 8441426496 utils.py:240] [process=0][thread=MainThread] Skipping global process sync, barrier name: Checkpointer:save.checkpoint_2
I1012 21:35:48.814014 8441426496 type_handlers.py:562] [process=0][thread=MainThread] Validating params (raw input=144, unique=72) in TensorStore KvStore: KvStore({
  'base': {
    'driver': 'file',
    'path': '/Users/ananyaebhartari/Documents/Paris/My_pirate/jaxpi/examples/allen_cahn/default/ckpt/checkpoint_2.orbax-checkpoint-tmp-0/',
  },
  'cache_pool': 'cache_pool#ocdbt',
  'config': {
    'compression': {'id': 'zstd'},
    'max_decoded_node_bytes': 100000000,
    'max_inline_value_bytes': 1024,
    'uuid': 'a9ae1caf9cc636dec7608c0cfbfbb469',
    'version_tree_arity_log2': 4,
  },
  'context': {
    'cache_pool#ocdbt': {'total_bytes_limit': 100000000},
    'data_copy_concurrency': {},
    'file_io_concurrency': {'limit': 128},
    'file_io_sync': True,
    'ocdbt_coordinator': {},
  },
  'driver': 'ocdbt',
  'experimental_read_coalescing_interval': '1ms',
  'experimental_read_coalescing_merged_bytes': 500000000000,
  'experimental_read_coalescing_threshold_bytes': 1000000,
}).
I1012 21:35:48.814340 8441426496 atomicity.py:316] Renaming /Users/ananyaebhartari/Documents/Paris/My_pirate/jaxpi/examples/allen_cahn/default/ckpt/checkpoint_2.orbax-checkpoint-tmp-0 to /Users/ananyaebhartari/Documents/Paris/My_pirate/jaxpi/examples/allen_cahn/default/ckpt/checkpoint_2
I1012 21:35:48.814490 8441426496 checkpoint.py:174] Read CheckpointMetadata=CheckpointMetadata(init_timestamp_nsecs=1728783348781540000, commit_timestamp_nsecs=None) from /Users/ananyaebhartari/Documents/Paris/My_pirate/jaxpi/examples/allen_cahn/default/ckpt/checkpoint_2.orbax-checkpoint-tmp-0
I1012 21:35:48.814660 8441426496 checkpoint.py:191] Updated CheckpointMetadata=CheckpointMetadata(init_timestamp_nsecs=1728783348781540000, commit_timestamp_nsecs=1728783348814391000) to /Users/ananyaebhartari/Documents/Paris/My_pirate/jaxpi/examples/allen_cahn/default/ckpt/checkpoint_2.orbax-checkpoint-tmp-0
I1012 21:35:48.814822 8441426496 atomicity.py:518] [process=0][thread=MainThread] Finished saving checkpoint (finalized tmp dir) to `/Users/ananyaebhartari/Documents/Paris/My_pirate/jaxpi/examples/allen_cahn/default/ckpt/checkpoint_2`.
I1012 21:35:48.814864 8441426496 utils.py:240] [process=0][thread=MainThread] Skipping global process sync, barrier name: Checkpointer:finalize.checkpoint_2
Traceback (most recent call last):
  File "/Users/ananyaebhartari/Documents/Paris/My_pirate/jaxpi/examples/allen_cahn/main.py", line 40, in <module>
    app.run(main)
  File "/Users/ananyaebhartari/Documents/Paris/My_pirate/.venv1/lib/python3.11/site-packages/absl/app.py", line 308, in run
    _run_main(main, args)
  File "/Users/ananyaebhartari/Documents/Paris/My_pirate/.venv1/lib/python3.11/site-packages/absl/app.py", line 254, in _run_main
    sys.exit(main(argv))
             ^^^^^^^^^^
  File "/Users/ananyaebhartari/Documents/Paris/My_pirate/jaxpi/examples/allen_cahn/main.py", line 32, in main
    train.train_and_evaluate(FLAGS.config, FLAGS.workdir)
  File "/Users/ananyaebhartari/Documents/Paris/My_pirate/jaxpi/examples/allen_cahn/train.py", line 134, in train_and_evaluate
    serializable_layer_grad_direction.append({
                                             ^
  File "/Users/ananyaebhartari/Documents/Paris/My_pirate/jaxpi/examples/allen_cahn/train.py", line 135, in <dictcomp>
    layer: numpy_callback(direction) if direction is not None else None
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ananyaebhartari/Documents/Paris/My_pirate/jaxpi/examples/allen_cahn/train.py", line 131, in numpy_callback
    return jax.pure_callback(jnp.asarray, result_shape, x)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ananyaebhartari/Documents/Paris/My_pirate/.venv1/lib/python3.11/site-packages/jax/_src/callback.py", line 326, in pure_callback
    out_flat = pure_callback_p.bind(
               ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ananyaebhartari/Documents/Paris/My_pirate/.venv1/lib/python3.11/site-packages/jax/_src/core.py", line 439, in bind
    return self.bind_with_trace(find_top_trace(args), args, params)
                                ^^^^^^^^^^^^^^^^^^^^
  File "/Users/ananyaebhartari/Documents/Paris/My_pirate/.venv1/lib/python3.11/site-packages/jax/_src/core.py", line 1385, in find_top_trace
    top_tracer._assert_live()
  File "/Users/ananyaebhartari/Documents/Paris/My_pirate/.venv1/lib/python3.11/site-packages/jax/_src/interpreters/partial_eval.py", line 1741, in _assert_live
    raise core.escaped_tracer_error(self, None)
jax.errors.UnexpectedTracerError: Encountered an unexpected tracer. A function transformed by JAX had a side effect, allowing for a reference to an intermediate value with type float64[] wrapped in a DynamicJaxprTracer to escape the scope of the transformation.
JAX transformations require that functions explicitly return their outputs, and disallow saving intermediate values to global state.
The function being traced when the value leaked was loss at /Users/ananyaebhartari/Documents/Paris/My_pirate/.venv1/lib/python3.11/site-packages/jaxpi/models.py:143 traced for jit.
------------------------------
The leaked intermediate value was created on line /Users/ananyaebhartari/Documents/Paris/My_pirate/jaxpi/examples/allen_cahn/models.py:85:29 (AllenCahn.losses).
------------------------------
When the value was created, the final 5 stack frames (most recent last) excluding JAX-internal frames were:
------------------------------
/Users/ananyaebhartari/Documents/Paris/My_pirate/jaxpi/examples/allen_cahn/main.py:32:8 (main)
/Users/ananyaebhartari/Documents/Paris/My_pirate/jaxpi/examples/allen_cahn/train.py:101:22 (train_and_evaluate)
/Users/ananyaebhartari/Documents/Paris/My_pirate/.venv1/lib/python3.11/site-packages/jaxpi/models.py:195:16 (PINN.step)
/Users/ananyaebhartari/Documents/Paris/My_pirate/.venv1/lib/python3.11/site-packages/jaxpi/models.py:146:17 (PINN.loss)
/Users/ananyaebhartari/Documents/Paris/My_pirate/jaxpi/examples/allen_cahn/models.py:85:29 (AllenCahn.losses)
------------------------------

To catch the leak earlier, try setting the environment variable JAX_CHECK_TRACER_LEAKS or using the `jax.checking_leaks` context manager.
See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.UnexpectedTracerError
